sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reshuffle.java

```diff
@@ -81,28 +88,107 @@ public PCollection<KV<K, V>> expand(PCollection<KV<K, V>> input) {
                     }
                   }
                 }))
-        .apply("RestoreOriginalTimestamps", ReifyTimestamps.extractFromValues());
+        .apply(
+            "RestoreOriginalWindows",
+            Window.into(new RestoreWindowsFn<>(originalStrategy.getWindowFn())))
+        .apply("RestoreOriginalTimestamps", new RestoreTimestamps<>())
+        // Set the windowing strategy directly, so that it doesn't get counted as the user having
+        // set allowed lateness.
+        .setWindowingStrategyInternal(originalStrategy);
+  }
+
+  private static class RestoreWindowsFn<K, V, W extends BoundedWindow>
+      extends NonMergingWindowFn<KV<K, ValueInSingleWindow<V>>, W> {
+
+    private final WindowFn<?, W> originalWindowFn;
+
+    private RestoreWindowsFn(WindowFn<?, W> originalWindowFn) {
+      this.originalWindowFn = originalWindowFn;
+    }
+
+    @Override
+    public Collection<W> assignWindows(AssignContext c) throws Exception {
+      return Collections.singleton((W) c.element().getValue().getWindow());
+    }
+
+    @Override
+    public boolean isCompatible(WindowFn<?, ?> other) {
+      throw new UnsupportedOperationException(
+          String.format(
+              "%s.isCompatible() should never be called."
+                  + " It is a private implementation detail of sdk utilities."
+                  + " This message indicates a bug in the Beam SDK.",
+              getClass().getCanonicalName()));
+    }
+
+    @Override
+    public void verifyCompatibility(WindowFn<?, ?> other) throws IncompatibleWindowException {
+      throw new UnsupportedOperationException(
+          String.format(
+              "%s.verifyCompatibility() should never be called."
+                  + " It is a private implementation detail of sdk utilities."
+                  + " This message indicates a bug in the Beam SDK.",
+              getClass().getCanonicalName()));
+    }
+
+    @Override
+    public Coder<W> windowCoder() {
+      return originalWindowFn.windowCoder();
+    }
+
+    @Override
+    public WindowMappingFn<W> getDefaultWindowMappingFn() {
+      throw new UnsupportedOperationException(
+          String.format(
+              "%s.getSideInputWindow() should never be called."
+                  + " It is a private implementation detail of sdk utilities."
+                  + " This message indicates a bug in the Beam SDK.",
+              getClass().getCanonicalName()));
+    }
+  }
+
+  private static class RestoreTimestamps<K, V>
+      extends PTransform<PCollection<KV<K, ValueInSingleWindow<V>>>, PCollection<KV<K, V>>> {
+    @Override
+    public PCollection<KV<K, V>> expand(PCollection<KV<K, ValueInSingleWindow<V>>> input) {
+      return input.apply(
+          ParDo.of(
+              new DoFn<KV<K, ValueInSingleWindow<V>>, KV<K, V>>() {
+                @Override
+                public Duration getAllowedTimestampSkew() {
+                  return Duration.millis(Long.MAX_VALUE);
```

Comment:
The runner doesn't read/act on this value anywhere, does it?

Reply:
Noting that this is not a change but a move. But no the runner does not act on it. In fact it is unsafe for this reason - the runner may well drop data that you have allowed to be output.